source/type/headsign_handler.cpp

```diff
@@ -0,0 +1,169 @@
+    // if no headsign map for vj: return name
+    if (!navitia::contains(map_vj_map_stop_time_headsign_change, stop_time.vehicle_journey)) {
+        return stop_time.vehicle_journey->name;
+    }
+
+    // otherwise use headsign change map
+    const auto& map_stop_time_headsign_change =
+            map_vj_map_stop_time_headsign_change.at(stop_time.vehicle_journey);
+    uint16_t order_stop_time = stop_time.journey_pattern_point->order;
+
+    // if no headsign change stored: return name
+    if (map_stop_time_headsign_change.empty()) {
+        return stop_time.vehicle_journey->name;
+    }
+
+    // get next change
+    auto it_headsign = map_stop_time_headsign_change.upper_bound(order_stop_time);
+    // if next change is the first: return name
+    if(it_headsign == map_stop_time_headsign_change.begin()) {
+        return stop_time.vehicle_journey->name;
+    }
+
+    // get previous change and return headsign
+    if (it_headsign == map_stop_time_headsign_change.end()) {
+        return map_stop_time_headsign_change.rbegin()->second;
+    }
+    return (--it_headsign)->second;
+}
+
+bool HeadsignHandler::has_headsign_or_name(const VehicleJourney& vj,
+                                           const std::string& headsign) const {
+    if (vj.name == headsign) {
+        return true;
+    }
+
+    if(!navitia::contains(map_vj_map_stop_time_headsign_change, &vj)) {
+        return false;
+    }
+
+    for (const auto& change: map_vj_map_stop_time_headsign_change.at(&vj)) {
+        if (change.second == headsign) {
+            return true;
+        }
+    }
+    return false;
+}
+
+std::vector<const VehicleJourney*>
+HeadsignHandler::get_vj_from_headsign(const std::string& headsign) const {
+    using namespace std;
+    vector<const VehicleJourney*> res;
+    const auto& it_vj_set = headsign_mvj.find(headsign);
+    if (it_vj_set == headsign_mvj.end()) {
+        return res;
+    }
+
+    for (const MetaVehicleJourney* mvj: it_vj_set->second) {
+        vector<vector<VehicleJourney*>> vect_vect_vj{mvj->theoric_vj, mvj->adapted_vj, mvj->real_time_vj};
+        for (const auto vect_vj: vect_vect_vj) {
+            for (const VehicleJourney* vj: vect_vj) {
+                if (has_headsign_or_name(*vj, headsign)) {
+                    res.push_back(vj);
+                }
+            }
+        }
+    }
+    return res;
+}
+
+void HeadsignHandler::affect_headsign_to_stop_time(const StopTime& stop_time,
+                                                   const std::string& headsign) {
+    const VehicleJourney* vj = stop_time.vehicle_journey;
+    headsign_mvj[vj->name].insert(vj->meta_vj); // for safety
+    std::string prev_headsign_for_stop_time = get_headsign(stop_time);
+    if (headsign == prev_headsign_for_stop_time) {
+        return;
+    }
+
+    uint16_t order = stop_time.journey_pattern_point->order;
+    // erase change if exists
+    if (navitia::contains(map_vj_map_stop_time_headsign_change[vj], order)) {
```

Comment:
I find this name a bit too long. 

Not sure but would `heasign_changes` be better ?